### PR TITLE
fix: handle 404 on cloud functions

### PR DESCRIPTION
**Pull request by mirestrepo**
This one was a bit tricky to figure out so I want to write out what I learned. The solution may not be perfect or the best, but it's the one way I got it to work. The issue arise from the combination of the way Nuxt handles 404,  and the fact that our website is not fully static

### If our site was fully static

We would need Firebase to redirect to a 404.html. To accomplish this one, would need to tell Nuxt to generate the `404.html` page in the config (we do that) by adding to the nuxt config file:

```
  generate: {
    fallback: "404.html"
  },
```

and to add `"destination": "/404.html"` to the firebase config. We tried the following

```
      "rewrites": [
         {
           "source": "**",
           "function": "ssrapp",
           "destination": "/404.html"
         }
       ],
```

At first we thought that worked, but then realized that adding `"destination": "/404.html"` overshadows the function and the dynamic routes never get hit

**Links**
* [Nuxt issue thread](https://github.com/nuxt/nuxt.js/issues/966)
* [Firebase custom 404 page](https://firebase.google.com/docs/hosting/full-config#404)

### Final solution.

The solution is a bit inspired by [this post](https://school.geekwall.in/p/BeMpV4tF). More specifically by these comments:

> ### Breakdown of the problem
> 
> - Nuxt doesn’t give middlewares an easy way to see if a route is valid or not
> - Nuxt runs its own 404 handler that can’t be intercepted in the middleware chain
> - Nuxt doesn’t have a clearly documented way of overriding the 404 handler on the server side
> 
> ### What we need to do
> 
> - Intercept all traffic and check if the requests are renderable by Nuxt
> - Proxy all other requests

To make things more explicit.

- All static pages (the ones that come from Nuxt Contex) are hit right away. This is handled by firebase.
- When a route is not found in the static content (`/dist`) folder, then the cloud function is called. This is handled by firebase (config - `"function": "ssrapp",`)
- In the cloud function I am keeping a list of the dynamic routes that the Nuxt app should handle, every other route is redirected to the static `404.html`

If we don't do this and let Nuxt handle the non existing route, we get a 500 instead of  404, and the website becomes unresponsive unless refreshed. I think this is because Nuxt my not treat 404 as an error and therefore our `try/catch` doesn't handle it appropriately

### CODE REVIEW

File: functions/index.js [hunk 1]
```diff
@@ -19,11 +21,19 @@ async function handleRequest(req, res) {
     }
     console.log(req.path);
     res.set('Cache-Control', 'public, max-age=3600, s-maxage=7200');
-    await nuxt.server.app.handle(req, res, (out) => console.log(out));
+    if(dynamicRoutes.includes(req.path)){ 
+      await nuxt.server.app.handle(req, res);
+    }
+    else{
```
Comment (broarr) [inline functions/index.js:27]: I'd stick this on one line, not that it matters though...

File: functions/index.js [hunk 2]
```diff
@@ -11,6 +11,8 @@ let isReady = false;
 
 let nuxt = loadNuxt('start');
 
+let dynamicRoutes = ['/_ghapi/status', '/_workday/opportunities']
```
Comment (broarr) [inline functions/index.js:14]: I'd consider renaming this to whitelistedRoutes, that way we know that these routes are allowed by the cloud function. It's the normal language used for this sort of thing in networking. Obviously this doesn't matter either ;-)

File: functions/index.js [hunk 3]
```diff
@@ -19,11 +21,19 @@ async function handleRequest(req, res) {
...
     console.log(req.path);
     res.set('Cache-Control', 'public, max-age=3600, s-maxage=7200');
-    await nuxt.server.app.handle(req, res, (out) => console.log(out));
+    if(dynamicRoutes.includes(req.path)){ 
+      await nuxt.server.app.handle(req, res);
+    }
+    else{
+      console.log('Redirect to 404 page');
+      res.set('X-Cascade', 'PASS')
+      res.status(404).redirect('/404.html')
```
Comment (broarr) [inline functions/index.js:30]: I think y'all might need a style enforcer like `prettier` or `standard` I'm seeing `;` sometimes, but not others

File: server-middleware/gh-api.js [hunk 4]
```diff
@@ -66,6 +66,9 @@ app.all('/status', async (req, res, next) => {
 
   try {
     const response = await client.request(statusQuery('ccv-status', 'OPEN'));
+    // eslint-disable-next-line no-console
+    console.log('GH Response')
+    console.log(response);
```
Comment (broarr) [inline server-middleware/gh-api.js:71]: You can probably delete this now

File: functions/index.js [hunk 5]
```diff
@@ -19,11 +21,19 @@ async function handleRequest(req, res) {
     }
     console.log(req.path);
     res.set('Cache-Control', 'public, max-age=3600, s-maxage=7200');
-    await nuxt.server.app.handle(req, res, (out) => console.log(out));
+    if(dynamicRoutes.includes(req.path)){ 
+      await nuxt.server.app.handle(req, res);
+    }
+    else{
+      console.log('Redirect to 404 page');
```
Comment (broarr) [inline functions/index.js:28]: Do you need this log still?

**Review by broarr — APPROVED** [review]
This looks good to me! Couple of style comments, but nothing that should prevent a merge!

File: README.md [hunk 6]
```diff
@@ -1,8 +1,49 @@
...
+],
+```
+
+In order to handle our 404 page correctly, we control which routes are served dynamically by the Nuxt app. As of (7-22-2022), the only dynamic routes configured in [functions/index.js](functions/index.js) are:
+
+```
+let dynamicRoutes = ['/_ghapi/status', '/_workday/opportunities']
+```
+
+All other routes that are not included on that list or in the `dist` folder are redirected to a pre-generated `404.html`. By default, Nuxt does not generate a `404.html` (only a 404.vue), to get the html page, he added the following block in the config.
```
Comment (eldu) [inline README.md:28]: pronoun change?

```suggestion
All other routes that are not included on that list or in the `dist` folder are redirected to a pre-generated `404.html`. By default, Nuxt does not generate a `404.html` (only a 404.vue), to get the html page, we added the following block in the config.
```

File: README.md [hunk 7]
```diff
@@ -1,8 +1,49 @@
...
+  },
+```
+
+## GitHub Actions
+
+There are two main workflows:
+
+1. When a PR is opened against the main branch, we deploy to qa-ccv-brown-edu live and preview channels.
+   There is no such thing as "preview" functions, so if the changes include changes to the functions, one needs to check qa-ccv-brown-edu.web.app (main channel). If the changes are only to static pages, then the preview channel will reflec the changes appropiately.
+2. When a PR is merged to the main branch, we deploy to ccv-brown-edu (live channel and fuctions)
```
Comment (eldu) [inline README.md:42]: typo

```suggestion
2. When a PR is merged to the main branch, we deploy to ccv-brown-edu (live channel and functions)
```

File: README.md [hunk 8]
```diff
@@ -20,28 +61,30 @@ git submodule update --remote
...
 GITHUB_USER=
 GITHUB_TOKEN=
 ```
 
-For detailed explanation on how things work, check out [Nuxt.js docs](https://nuxtjs.org).
-
-
 ## Testing build set up
 
+In deployment, the dynamic routes are served by Firebase fuctions. To simulate this behavior, we need to run through the entire build process, which asks Nuxt to generate the static routes and firebase to emulate the functions
```
Comment (eldu) [inline README.md:73]: typo

```suggestion
In deployment, the dynamic routes are served by Firebase functions. To simulate this behavior, we need to run through the entire build process, which asks Nuxt to generate the static routes and firebase to emulate the functions
```

File: README.md [hunk 9]
```diff
@@ -20,28 +61,30 @@ git submodule update --remote
...
 npm run fn-install # if needed
 npm run local-deploy
 ```
 
-May need to set up github credentials with functions env:
-```
-firebase functions:config:set gh.token="THE API KEY" gh.user="THE CLIENT ID"
-```
+**Note:**
+* 2022-06-23 (MIR): Our local-deploy uses `firebase serve` which seems to not be recommended anymore. We'll probably need to run teh fuctions emulator instead.
```
Comment (eldu) [inline README.md:82]: typos

```suggestion
* 2022-06-23 (MIR): Our local-deploy uses `firebase serve` which seems to not be recommended anymore. We'll probably need to run the functions emulator instead. 
```

**Review by eldu — APPROVED** [review]
Thanks you for tackling this weird problem! I didn't realize how weirdly nuxt handles 404! Great learnings in here and thank you for documenting them as well as instructions for updates.

I made comments on some typos. And then Brad also mentioned changes as well. Looking very good Isabel! Thanks again!

File: functions/index.js (hunk 5 above)
Comment (mirestrepo) [reply functions/index.js:28]: Probably not

File: functions/index.js (hunk 3 above)
Comment (mirestrepo) [reply functions/index.js:30]: Probably… we have it in the parent dir… may just need to add it to the functions dir..

File: functions/index.js (hunk 2 above)
Comment (mirestrepo) [reply functions/index.js:14]: I don’t love the whitel/black list terminology, but I could use allowedRoutes or something more clear

**Comment by mirestrepo** [comment]
Thanks @broarr and @eldu for reviewing! I’ve merged the docs changes since I can do it from my ipad and made #107 to address Brad’s suggestions. I’m gonna merge so this bug doesn’t stay around longer

**Pull request merged**